Mathlib/Data/Nat/Squarefree.lean

```diff
@@ -384,6 +391,17 @@ theorem squarefree_mul_iff {m n : ℕ} :
   ⟨fun h => ⟨coprime_of_squarefree_mul h, (squarefree_mul $ coprime_of_squarefree_mul h).mp h⟩,
     fun h => (squarefree_mul h.1).mpr h.2⟩
 
+theorem prod_factors_toFinset_of_squarefree {n : ℕ} (hn : Squarefree n) :
+    ∏ p in n.factors.toFinset, p = n := by
+  erw [List.prod_toFinset _ hn.nodup_factors, List.map_id, Nat.prod_factors hn.ne_zero]
+
+theorem prod_factors_toFinset_sdiff_of_squarefree {n : ℕ} (hn : Squarefree n) {t : Finset ℕ}
+    (ht : t ⊆ n.factors.toFinset) :
+    ∏ a in (n.factors.toFinset \ t), a = n / ∏ a in t, a := by
+  refine symm $ Nat.div_eq_of_eq_mul_left (Finset.prod_pos
+    fun p hp => (prime_of_mem_factors (List.mem_toFinset.mp (ht hp))).pos) ?_
+  rw [Finset.prod_sdiff ht, prod_factors_toFinset_of_squarefree hn]
```

Comment:
The other two theorems are fine but this one strikes me as a bit odd. Firstly, any theorem about division of the form X = Y / Z when Z actually divides Y is strictly weaker than the corresponding statement about multiplication X * Z = Y, because natural division is a pathological operation (e.g. 5/2=2). So surely it's better stated as a lemma about multiplication? And then it basically becomes `Finset.prod_sdiff`. Are you sure you need this theorem? I would avoid natural division and subtraction if at all possible, rather than proving lemmas about them.

Reply:
I agree that this looks dubious, but I think it might take quite a lot of effort to properly avoid. Roughly speaking natural division comes up when working with Dirichlet convolution and in my case is difficult to avoid because `Nat.sum_divisors_filter_squarefree` only really makes sense for a sum over `divisors` instead of `divisorsAntidiagonal`.

I actually originally extracted this as a rewrite lemma because both sides came up nested in subexpressions, but I could just call `congr` and not make it a lemma. Would that be preferable?

To give a little more detail on why natural division comes up, I want to prove that for $f, g$ multiplicative arithmetic functions and squarefree $n$ we have $\prod_{p\mid n} (f+g)(p) = (f*g) (n)$. In the proof, I write both sides as a sum over subsets of `factors n`. Expanding the product then naturally gives the LHS of the lemma. Writing the convolution as $\sum _{d\mid n} f(d) g(n/d)$ and applying `Nat.sum_divisors_filter_squarefree` produces the expression on the RHS.

Reply:
If you are convinced you need a lemma with division then this is an argument for having the lemma. I am not convinced yet but I certainly remain open to becoming convinced. 

I agree that `Nat.sum_divisors_filter_squarefree` should be a sum over `divisors`. I agree that *if* you write the convolution as a sum involving division (which my instinct is where the code smell is) as opposed to `divisorsAntidiagonal` then you will have to deal with divisions. Because I haven't seen the explicit code I am currently unclear about why you think that it's helpful to write the convolution using division rather than using `divisorsAntidiagonal`. Could it be that you're missing a trick involving manipulating finite sums over different sets?

Look at this this way. Division is a pathological operation, so the moment you have a term `a / b` you are essentially unable to do anything at all with it until you have multiplied by b again, proved that actually b did divide a, and then applied some div_mul_cancel lemma. You are saying "I need division" but I am saying "my mental model is that division cannot ever help because it basically freezes part of the code until you undivide, so why not just delay the division until you discover that it's gone". 

Can you link to some code where you think that division is making your life easier? Then this PR would make a whole lot more sense to me.

Reply:
[Here's the proof](https://github.com/FLDutchmann/selberg-sieve4/blob/38c9db4fdb2784a09113c04df4e6195196e1a6d0/SelbergSieve/ForArithmeticFunction.lean#L78), for reference. I claim that `Nat.sum_divisors_filter_squarefree` makes the proof much simpler by lining up the sums perfectly along the powerset of `factors n`.

Reply:
> difficult to avoid because `Nat.sum_divisors_filter_squarefree` only really makes sense for a sum over `divisors` instead of `divisorsAntidiagonal`.

I think you can combine `image_fst_divisorsAntidiagonal` with `Nat.sum_divisors_filter_squarefree` to obtain a sum over `divisorsAntidiagonal` and apply Kevin's strategy from there.